File: leetcode/Easy/python/108. Convert Sorted Array to Binary Search Tree.py
'''
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/

Решение задачи "Convert Sorted Array to Binary Search Tree".
'''

# Definition for a binary tree node.
# В Python класс TreeNode уже объявлен LeetCode.

class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None

        mid = len(nums) // 2
        root = TreeNode(nums[mid])

        root.left  = self.sortedArrayToBST(nums[:mid])
        root.right = self.sortedArrayToBST(nums[mid+1:])

        return root

File: leetcode/Easy/java/108. Convert Sorted Array to Binary Search Tree.java
/*
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/

Решение задачи "Convert Sorted Array to Binary Search Tree".
*/

class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return build(nums, 0, nums.length - 1);
    }

    private TreeNode build(int[] nums, int left, int right) {
        if (left > right) return null;

        int mid = left + (right - left) / 2;
        TreeNode node = new TreeNode(nums[mid]);

        node.left  = build(nums, left, mid - 1);
        node.right = build(nums, mid + 1, right);

        return node;
    }
}
